RFC-0020-Lightweight-Dispatch.md

```diff
@@ -0,0 +1,287 @@
+* For the scenario of adding a new operator (not to `native_functions.yaml`), we need to provide clear guidance to add it to the JIT op registry as well, otherwise JIT execution will break.
+* We can add tests on the mobile build for the sake of coverage.
+
+For OSS mobile integration, we will need to have a new build flavor to switch between c10 dispatcher vs jit op registry. This new flavor will include codegen source files (`UnboxingFunctions.h, UnboxingFunctions.cpp, RegisterCodegenUnboxedKernels.cpp`) instead of existing dispatcher related source files: `Operators.cpp`, `RegisterSchema.cpp `etc, similar to the internal build configuration. Again, this will be delivered as a build flavor for user to opt in, the dispatcher will be used by default.
+
+
+
+### Step 2
+
+With step 1 we already have a working codegen unboxing + static dispatch system working but it only works for the `CPU` backend. Nowadays most models being deployed on edge devices are quantized models so we will need to support both `CPU` and `QuantizedCPU` backend. In addition to that, a lot of our models feature custom ops, however we can’t register custom ops through the old dispatcher  (`TORCH_LIBRARY`) APIs any more. Here I’m proposing a solution that exposes the `native_function.yaml` syntax to the internal developers targeting this runtime mode: allow them to use the yaml file format to declare their custom ops and/or custom kernels.
+
+ 
+
+
+#### Support multiple backends in static dispatch
+
+**NOTE: this may be optional if we enabled backend tracing for ops.** For the vast majority of models, we will only have 1 backend per operator, meaning that if we can pass the backend info into codegen, we don’t have to do dispatch based on dispatch key.
+
+In the scenario that a model contains both floating point ops and quantized ops, our codegen should be able to statically dispatch to the correct backend. The following diagram shows what will be generated and included in the build and demonstrates the dependency relationship.
+
+Let’s take `acosh` as an example:
+```yaml
+native_functions.yaml
+=====================
+- func: acosh(Tensor self) -> Tensor
+  variants: function, method
+  structured_delegate: acosh.out
+
+- func: acosh.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
+  structured: True
+  structured_inherits: TensorIteratorBase
+  dispatch:
+    CPU, CUDA: acosh_out
+```
+
+And if we pass the backends we want to codegen for both `CPU` and `QuantizedCPU` backends, our `Functions.h` will be something like this (borrowing from Jiakai’s [PR](https://github.com/pytorch/pytorch/pull/51554/commits/0ba3d4cc42187f69f17e0c382f0ab51e071a4a44)):
+
+```cpp
+Functions.h
+===========
+// aten::acosh(Tensor self) -> Tensor
+TORCH_API inline at::Tensor acosh(const at::Tensor & self) {
+    DispatchKeySet _dk_set = c10::detail::multi_dispatch_key_set(tensor);
+    DispatchKey _dk = _dk_set.highestPriorityBackendTypeId();
+    switch (_dk) {
+    case DispatchKey::CPU:
```

Comment:
If it's runtime-control flow, does it mean that we need to build both the kernels, even if we just use one? We could use tracing based selective build, but tracing would still be required.

Reply:
We should trace backend AOT to avoid building both kernels. I think Dhruv has a proposal of doing so.